comp/forwarder/defaultforwarder/blocked_endpoints.go

```diff
@@ -67,15 +121,48 @@ func (e *blockedEndpoints) close(endpoint string) {
 	if knownBlock, ok := e.errorPerEndpoint[endpoint]; ok {
 		b = knownBlock
 	} else {
-		b = &block{}
+		b = &block{
+			endpoint: endpoint,
+			state:    unblocked,
+		}
 	}
 
-	b.nbError = e.backoffPolicy.IncError(b.nbError)
-	b.until = time.Now().Add(e.getBackoffDuration(b.nbError))
+	b.m.RLock()
+
+	switch b.state {
+	case unblocked:
+		// The circuit breaker is unblocked, we need to block for a time determined by the
+		// backoff policy.
+		b.m.RUnlock()
+		b.m.Lock()
+		defer b.m.Unlock()
+
+		if b.state == unblocked {
```

Comment:
Since we have write lock on `e.m`, `b.state` can not change from line 132 to here, and this check is redundant.